File: Searching/Cyclic Sort/CyclicSort.cpp
#include <bits/stdc++.h>
#include<iostream>

using namespace std;

class Solution{
    public:
    void swap(vector<int> arr, int i, int correct){
        int temp = arr[i];
        arr[i] = arr[correct];
        arr[correct] = temp;
    }
    void sort(vector<int> nums){
        int i = 0;
        while (i < nums.size()){
            int correct = nums[i] - 1;
            if (nums[i] != nums[correct]) swap(nums, i, correct);
            else i++;
        }
    }
};

 main(){
    Solution s;
    vector<int> arr = {4, 5, 2, 3, 1};
    s.sort(arr);
    
    for(int i = 0; i < arr.size(); i++) cout<< arr[i] << " ";
}

File: Searching/Cyclic Sort/CyclicSort.py
from typing import List

class Solution:
    def sort(self, nums: List[int]) -> List[int]:
        i = 0
        count = 0
        while i < len(nums):
            correct = nums[i] - 1
            if nums[i] != nums[correct]:
                nums[i], nums[correct] = nums[correct], nums[i]
                count += 1
            else:
                i += 1
        print(count)
        return nums

def main():
    s = Solution()
    nums = [5, 4, 6, 3, 2, 1]
    print(s.sort([2,2,1,2]))

main()